src/include/cart/api.h

```diff
@@ -685,98 +685,152 @@ crt_ep_abort(crt_endpoint_t *ep);
  # CRT_GEN_PROC_FUNC(struct, CRT_SEQ_MY_TYPE)
  *
  */
-#define CRT_VAR   0
-#define CRT_PTR   1
-#define CRT_ARRAY 2
-#define CRT_RAW   3
+#define CRT_VAR		(0)
+#define CRT_PTR		(1)
+#define CRT_ARRAY	(2)
+#define CRT_RAW		(3)
 
-#define CRT_GEN_GET_TYPE(seq) BOOST_PP_SEQ_HEAD(seq)
+#define CRT_GEN_GET_TYPE(seq) BOOST_PP_SEQ_ELEM(0, seq)
 #define CRT_GEN_GET_NAME(seq) BOOST_PP_SEQ_ELEM(1, seq)
-#define CRT_GEN_GET_KIND(seq) BOOST_PP_SEQ_TAIL(BOOST_PP_SEQ_TAIL(seq))
+#define CRT_GEN_GET_KIND(seq) BOOST_PP_SEQ_ELEM(2, seq)
 
-#define CRT_GEN_STRUCT_FIELD(r, data, seq)				\
-	BOOST_PP_IF(BOOST_PP_EQUAL(CRT_ARRAY, CRT_GEN_GET_KIND(seq)),	\
+/* convert constructed name into proper name */
+#define crt_proc_struct BOOST_PP_RPAREN() BOOST_PP_CAT BOOST_PP_LPAREN() \
+	crt_proc_struct_,
+
+#define CRT_GEN_X(x) x
+#define CRT_GEN_X2(x) CRT_GEN_X BOOST_PP_LPAREN() crt_proc_##x BOOST_PP_RPAREN()
+#define CRT_GEN_GET_FUNC(seq) CRT_GEN_X2 BOOST_PP_SEQ_FIRST_N(1, seq)
+#define CRT_GEN_X3(x) CRT_GEN_X BOOST_PP_LPAREN() fail_##x BOOST_PP_RPAREN()
+#define CRT_GEN_FAIL_LABEL(seq) CRT_GEN_X3 BOOST_PP_SEQ_TAIL(BOOST_PP_SEQ_FIRST_N(2, seq))
```

Comment:
(style)  line over 80 characters
(style)  Macros with complex values should be enclosed in parentheses